File: Data Structures/#04.Linear Data Structures - Stack And Queue Exercises/C#/T03_Array_Based_Stack/ArrayStack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

public class ArrayStack<T>
{
    private const int InitialCapacity = 16;

    private T[] elements;
    private int count;
    private int capacity;

    public int Count { get => count; set => count = value; }
    public int Capacity { get => capacity; set => capacity = value; }

    public ArrayStack(int capacity = InitialCapacity)
    {
        this.capacity = capacity;
        this.count = 0;
        this.elements = new T[this.capacity];
    }
    public void Push(T element)
    {

        if (this.count > this.capacity - 1)
        {
            this.Grow();
        }

        this.elements[this.count] = element;
        count++;
    }

    public T Pop()
    {
        if (count == 0)
        {
            throw new InvalidOperationException();
        }

        count--;
        T element = this.elements[this.count];
        this.elements[this.count] = default(T);

        return element;
    }
    public T[] ToArray()
    {
        T[] resultedArray = new T[this.count];
        this.CopyAllElementsTo(ref resultedArray);
        this.Reverse(ref resultedArray);
        return resultedArray;
    }
    private void Grow()
    {
        this.capacity *= 2;
        T[] newArray = new T[this.capacity];
        this.CopyAllElementsTo(ref newArray);
        this.elements = newArray;

    }

    private void Reverse(ref T[] array)
    {
        T[] newArr = new T[array.Length];
        int counter = 0;
        for (int i = array.Length - 1; i >= 0; i--)
        {
            newArr[counter] = array[i];
            counter++;
        }

        array = newArr;
    }

    private void CopyAllElementsTo(ref T[] array)
    {

        for (int i = 0; i < this.count; i++)
        {
            array[i] = this.elements[i];
        }

    }
}

File: Data Structures/#04.Linear Data Structures - Stack And Queue Exercises/Java/src/T03_Array_Based_Stack/ArrayStack.java
package T03_Array_Based_Stack;

public class ArrayStack<T> {

    private static final int INITIAL_CAPACITY = 16;
    private static final int INITIAL_SIZE = 0;

    private T[] elements;
    private int size;
    private int count;

    public ArrayStack() {
        this.count = INITIAL_SIZE;
        this.size = INITIAL_CAPACITY;
    }

    public ArrayStack(int capacity) {
        this();
        this.size = capacity;
    }

    public int size() {
        return this.size;
    }

    private void setSize(int size) {
        this.size = size;
    }

    public void push(T element) {

        if (this.count == 0) {
            this.elements = (T[]) new Object[this.size];
            this.elements[this.count] = element;
            this.count++;
            return;
        }

        if (this.count == this.size) {
            this.grow();
        }

        this.elements[this.count] = element;
        this.count++;
    }

    public T pop() {

        if(this.count-1 < 0){
            throw new IllegalArgumentException();
        }

        T element = this.elements[this.count-1];
        this.elements[this.count-1] = null;

        return element;
    }

    public T[] toArray() {
        T[] resultArray = (T[]) new Object[this.count];

        this.fillArray(resultArray);

        return resultArray;
    }

    private void grow() {
        this.size = this.size * 2;
        T[] newArray = (T[]) new Object[this.size];

        this.fillArray(newArray);

        this.elements = newArray;
    }

    private void fillArray(T[] newArray){
        for (int i = 0; i < this.count; i++) {
            newArray[i] = this.elements[i];
        }
    }

}